Serva w4 javascript

Wandering-Souls opened the pull request:
Serveh Saeedi - JavaScript Homework W4

Erlemorgaine commented:
Hi Serva! Overall your code looks good, so well done!

One general remark: I think in the README guide it says that you guys have to remove the comments starting with 'TODO' from the code once you're done, and this will also clean up the files a little. Can you do this for each exercise?

Another general remark: in some exercises you added console.log statements. This is perfectly fine to test and debug your code when you're working on it, but once you're submitting it for review, it's better to remove them, to keep the code more clean and (once you're working as a developer) not log things in production code that shouldn't be visible in the browser for the client. Could you do this?

### Exercise 1

* Your solution looks good! There is one thing that you can do to make the code in your function one line smaller (and still reasonably readable). Basically there's one thing that you're doing now that you don't necessarily need to do, and without which the code would function as well. Can you guess what that would be?

### Exercise 2

* Also here, well done! I have just a small tip: in general it's good practice to make your variable names as specific as possible, so that it's clear for other developers what's going on in your code. For example, I think the name `taskDurations` is well chosen. However, the variable name `sum` is very generic, it could be a sum of anything. Could you make the name more specific?
(Also, remove the log statements, both from your code and from the test :) )

### Exercise 3

* It's a good solution, and your variable names are specific, well done! When I run your tests here, I still notice a lint error (`'includes' is assigned a value but never used`). Could you change the code so that the error does't show?

### Exercise 4 and 5

Nothing to add here, well done!

Wandering-Souls commented:
Hello Erle, thanks very much for reviewing my homework and all the tips and suggestions for improving my codes. 

I removed all the console.log statements and TODO's. Good tip about console.log, at the moment I just think about solving my exercise but it's better to get used to think about client side as well. Thank you for the tip. 

Regarding exercise 1, I've removed "const" variable and put my statement directly in the return so now I have one line less codes. I hope this is what exactly you mean. 

In exercise 2, I've removed extra logs and renamed the "sum" to "totalEarning". 

Regarding exercise 3, I removed "const" variable and put my statement directly in the return just like ex-1 but I really don't understand the error about includes. I removed "allergenFruit" which is a value inside includes and I received an error. Also I substituted it with lemon and I got the same error so I don't know what to change exactly. 

About the second test in ex-3, I didn't get your point really, should I change something? 
we've given a variable originalFruitBasketContents in the test so I've compared that with fruitBasket. Based on your comment I called sanitizeFruitBasket and I got an error.  Could you please clarifye that for me? 

Thanks again for correction!

Wandering-Souls commented:
I remove the weird includes statement from my codes!

Erlemorgaine commented:
Hi Serva,

All changes approved. The only other thing that you need to do is call `sanitizeFruitBasket` in **test 2 of exercise 3**, as we discussed last Monday. If something is still not clear here or if you still get an error when you try to call the function, let me know.

One other thing that I forgot to mention when I reviewed the first time: I liked that you use `includes` here:

```
return fruits.filter((fruit) => !fruit.includes(allergenFruit));
```

but I think for this exercise it's enough, and possibly safer, to just check if `fruit` and `allergenFruit` are not equal to each other.

It's important to understand what `includes` does here:
`fruit` is a string. When you call `includes` on a string, it will check if a substring is part of that string.  I'll give some examples. 

In this first example, you check if the string 'lemon' is part of the string 'lemon'. This is, of course, true, since the strings are equal:

```
const fruit = 'lemon';

fruit.includes('lemon');  // Result: true

```

Here, you're checking if 'lemon' is a substring of the string 'slice of lemon'. And indeed it is. This is why I like your example: it also filters slices of lemon out of the fruit basket :)
```
const fruit = 'slice of lemon';

fruit.includes('lemon');   // Result: true
```

Here is an example of where things might go wrong. Suppose you also put some lemonade in your fruit basket (which would be a bit weird, since lemonade is not a fruit... But let's assume it's okay). Now the lemonade will also be removed from your fruit basket, since the word 'lemonade' contains the substring 'lemon'. And probably we don't want that.
```
const fruit = 'lemonade';

fruit.includes('lemon');   // Result: true
```

All this is to say: I like your example, since it would also filter out other things that contain the fruit that you're allergic to. But, it might also filter out unwanted things. So it's probably safer to stick to checking if `fruit` and `allergenFruit` are not equal to each other.

Hope all this is clear, if not let me know!

Wandering-Souls commented:
Hi Erle, thank you so much for your detailed explanation. Actually I had already changed test 2 of exercise 3. My comment is for last week after I saw your review. Anyway, I have replaced includes in the solution. Hope this time everything is ok. 
Thanks again!
